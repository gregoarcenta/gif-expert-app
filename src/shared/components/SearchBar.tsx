import { useEffect, useState } from "react";

interface SearchBarProps {
  placeholder?: string;
  currentQuery: string;
  onQuery: (query: string) => void;
}

function SearchBar({
  placeholder = "Buscar gifs",
  currentQuery,
  onQuery,
}: SearchBarProps) {
  const [query, setQuery] = useState(currentQuery);

  useEffect(() => {
    setQuery(currentQuery);
  }, [currentQuery]);

  useEffect(() => {
    const term = query.toLowerCase().trim();
    if (term.length === 0) return;

    const timeoutId = setTimeout(() => {
      onQuery(term);
    }, 700);

    return () => {
      clearTimeout(timeoutId);
    };
  }, [onQuery, query]);

  const handleSearch = () => {
    onQuery(query);
    setQuery("");
  };

  return (
    <div className="search-container">
      <input
        type="text"
        placeholder={placeholder}
        value={query}
        onChange={(e) => setQuery(e.target.value)}
        onKeyDown={(e) => e.key === "Enter" && handleSearch()}
      />
      <button onClick={handleSearch}>Buscar</button>
    </div>
  );
}

export default SearchBar;
