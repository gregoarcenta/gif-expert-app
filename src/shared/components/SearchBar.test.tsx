import { describe, expect, test, vi } from "vitest";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import SearchBar from "./SearchBar.tsx";

describe("SearchBar", () => {
  test("should render searchBar correctly", () => {
    const onQuery = vi.fn();
    const { container } = render(
      <SearchBar previousQuery={""} onQuery={onQuery} />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should call onQuery with the correct value after 700ms", async () => {
    const onQuery = vi.fn();
    render(<SearchBar previousQuery={""} onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "test" } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledExactlyOnceWith("test");
    });
  });

  test("should call only once with the last value (debounce)", async () => {
    const onQuery = vi.fn();
    render(<SearchBar previousQuery={""} onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "t" } });
    fireEvent.change(input, { target: { value: "te" } });
    fireEvent.change(input, { target: { value: "tes" } });
    fireEvent.change(input, { target: { value: "test" } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledExactlyOnceWith("test");
    });
  });

  test("should call onQuery when the button clicked with the input value", () => {
    const onQuery = vi.fn();
    render(<SearchBar previousQuery={""} onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "test" } });

    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(onQuery).toHaveBeenCalledExactlyOnceWith("test");
  });

  test("should the input has  the correct placeholder value", () => {
    const value = "Busca los mejores gifs";
    render(
      <SearchBar
        previousQuery={"previousQuery"}
        onQuery={() => {}}
        placeholder={value}
      />,
    );
    expect(screen.getByPlaceholderText(value)).toBeDefined();
  });

  test("should call to onQuery when previousQuery is not empty", async () => {
    const onQuery = vi.fn();
    const newQuery = "test previousQuery";

    const { rerender } = render(
      <SearchBar previousQuery="" onQuery={onQuery} />,
    );

    expect(onQuery).not.toHaveBeenCalled();

    rerender(<SearchBar previousQuery={newQuery} onQuery={onQuery} />);

    await waitFor(() =>
      expect(onQuery).toHaveBeenCalledExactlyOnceWith(newQuery.toLowerCase()),
    );
  });
});
