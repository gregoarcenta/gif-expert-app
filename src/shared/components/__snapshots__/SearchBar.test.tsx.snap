// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render searchBar correctly 1`] = `
<div>
  <div
    class="search-container"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
    <button>
      Buscar
    </button>
  </div>
</div>
`;
